### PR TITLE
Replacement of local cache in rule nodes with cached entity service APIs.

### PR DESCRIPTION
**Pull request by ShvaykaD**
## Pull Request description

1. Added cache for: 

 - UserServiceImpl.findUserByTenantIdAndEmail
 - CustomerServiceImpl.findCustomerByTenantIdAndTitleUsingCache

2. Removed guava cache for next rule nodes:

 - TbCreateRelationNode
 - TbDeleteRelationNode
 - TbAssignToCustomerNode
 - TbUnassignFromCustomerNode
  
image

3. Added CONSTRAINT customer_title_unq_key UNIQUE (tenant_id, title).

4. Added upgrade script to upgrade customers with same title.

PE PR: https://github.com/thingsboard/thingsboard-pe/pull/2467

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.
  
## Front-End feature checklist

- [x] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [x] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [x] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.
- [x] If new yml property was added: make sure a description is added (above or near the property).

### CODE REVIEW

File: dao/src/main/java/org/thingsboard/server/dao/customer/CustomerServiceImpl.java [hunk 1]
```diff
@@ -104,22 +131,34 @@ public ListenableFuture<Customer> findCustomerByIdAsync(TenantId tenantId, Custo
...
+        return saveCustomer(customer, true);
+    }
+
+    private Customer saveCustomer(Customer customer, boolean doValidate) {
         log.trace("Executing saveCustomer [{}]", customer);
-        customerValidator.validate(customer, Customer::getTenantId);
+        String oldCustomerTitle = null;
+        if (doValidate) {
+            Customer oldCustomer = customerValidator.validate(customer, Customer::getTenantId);
+            oldCustomerTitle = oldCustomer != null ? oldCustomer.getTitle() : null;
```
Comment (ShvaykaD) [inline dao/src/main/java/org/thingsboard/server/dao/customer/CustomerServiceImpl.java:142]: replace with single if

File: dao/src/main/java/org/thingsboard/server/dao/service/validator/CustomerDataValidator.java [hunk 2]
```diff
@@ -41,24 +39,15 @@ public class CustomerDataValidator extends DataValidator<Customer> {
...
-                c -> {
-                    if (!c.getId().equals(customer.getId())) {
-                        throw new DataValidationException("Customer with such title already exists!");
-                    }
-                }
-        );
-        return customerOpt.orElse(null);
+        Customer old = customerDao.findById(customer.getTenantId(), customer.getId().getId());
+        if (old == null) {
+            throw new DataValidationException("Can't update non existing customer!");
```
Comment (dskarzh) [inline dao/src/main/java/org/thingsboard/server/dao/service/validator/CustomerDataValidator.java:48]:
```suggestion
            throw new DataValidationException("Can't update nonexistent customer!");
```

File: dao/src/test/resources/application-test.properties [hunk 3]
```diff
@@ -26,6 +26,12 @@ cache.specs.sessions.maxSize=100000
 cache.specs.assets.timeToLiveInMinutes=1440
 cache.specs.assets.maxSize=100000
 
+cache.specs.customers.timeToLiveInMinutes=1440
+cache.specs.customers.maxSize=10000
+
+cache.specs.users.timeToLiveInMinutes=1440
+cache.specs.users.maxSize=10000
```
Comment (dskarzh) [inline dao/src/test/resources/application-test.properties:33]: other cache specs use TTL of 100000 (5 zeroes), customer cache specs uses TTL of 10000 (4 zeroes). Is this intentional?

File: common/cache/src/main/java/org/thingsboard/server/cache/customer/CustomerCacheEvictEvent.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import org.thingsboard.server.common.data.id.TenantId;
+
+@Data
+@RequiredArgsConstructor
+public
+class CustomerCacheEvictEvent {
+
+    private final TenantId tenantId;
+    private final String newTitle;
+    private final String oldTitle;
```
Comment (ShvaykaD) [inline common/cache/src/main/java/org/thingsboard/server/cache/customer/CustomerCacheEvictEvent.java:29]: Why we need to evict cache with newTitle if it's new and we have a UNIQUE constraint for title.

File: common/cache/src/main/java/org/thingsboard/server/cache/user/UserCacheEvictEvent.java [hunk 5]
```diff
@@ -13,16 +13,19 @@
...
-import org.thingsboard.server.common.data.EntityType;
-import org.thingsboard.server.common.data.id.EntityId;
+import lombok.RequiredArgsConstructor;
+import org.thingsboard.server.common.data.id.TenantId;
 
 @Data
-public class EntityContainer {
+@RequiredArgsConstructor
+public
+class UserCacheEvictEvent {
```
Comment (ShvaykaD) [inline common/cache/src/main/java/org/thingsboard/server/cache/user/UserCacheEvictEvent.java:25]:
```suggestion
public class UserCacheEvictEvent {
```

File: common/dao-api/src/main/java/org/thingsboard/server/dao/customer/CustomerService.java [hunk 6]
```diff
@@ -29,6 +29,9 @@ public interface CustomerService extends EntityDaoService {
 
     Customer findCustomerById(TenantId tenantId, CustomerId customerId);
 
+    Customer findCustomerByTenantIdAndTitleUsingCache(TenantId tenantId, String title);
+
+    @Deprecated(since = "3.6.4", forRemoval = true)
```
Comment (ShvaykaD) [inline common/dao-api/src/main/java/org/thingsboard/server/dao/customer/CustomerService.java:34]: fix on the impl side.

File: dao/src/main/java/org/thingsboard/server/dao/asset/BaseAssetService.java [hunk 7]
```diff
@@ -196,6 +196,9 @@ public Asset assignAssetToCustomer(TenantId tenantId, AssetId assetId, CustomerI
     @Override
     public Asset unassignAssetFromCustomer(TenantId tenantId, AssetId assetId) {
         Asset asset = findAssetById(tenantId, assetId);
+        if (asset.getCustomerId() == null) {
```
Comment (ShvaykaD) [inline dao/src/main/java/org/thingsboard/server/dao/asset/BaseAssetService.java:199]: do the same trick with assignAssetToCustomer

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/action/TbUnassignFromCustomerNode.java [hunk 8]
```diff
@@ -53,48 +60,39 @@ protected TbUnassignFromCustomerNodeConfiguration loadCustomerNodeActionConfig(T
...
+        var originatorType = originator.getEntityType();
+        var tenantId = ctx.getTenantId();
+        if (EntityType.DASHBOARD.equals(originatorType)) {
+            if (StringUtils.isEmpty(config.getCustomerNamePattern())) {
+                throw new RuntimeException("Failed to unassign dashboard with id '" +
+                        originator.getId() + "' from customer! Customer title should be specified!");
+            }
+            var customerIdFuture = getCustomerIdFuture(ctx, msg);
+            return Futures.transformAsync(customerIdFuture, customerId ->
+                    ctx.getDbCallbackExecutor().submit(() -> {
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/action/TbUnassignFromCustomerNode.java:74]: remove unnecessary use of   ctx.getDbCallbackExecutor().submit(() -> {

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/action/TbCreateRelationNode.java [hunk 9]
```diff
@@ -71,167 +84,56 @@ protected boolean createEntityIfNotExists() {
...
-                return processDashboard(ctx, entityContainer, sdId, relationType);
-            case ENTITY_VIEW:
-                return processView(ctx, entityContainer, sdId, relationType);
-            case EDGE:
-                return processEdge(ctx, entityContainer, sdId, relationType);
-            case TENANT:
-                return processTenant(ctx, entityContainer, sdId, relationType);
-            case USER:
-                return processUser(ctx, entityContainer, sdId, relationType);
+    private ListenableFuture<Void> deleteCurrentRelationsIfNeeded(TbContext ctx, EntityId originator, String relationType) {
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/action/TbCreateRelationNode.java:103]: move logic that is common for delete and create relation node to abstract class.

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/action/TbAbstractRelationActionNode.java [hunk 10]
```diff
@@ -63,222 +51,225 @@
...
-        }
-        entityIdCache = cacheBuilder.build(new EntityCacheLoader(ctx, createEntityIfNotExists()));
     }
 
     @Override
     public void onMsg(TbContext ctx, TbMsg msg) {
-        String relationType = processPattern(msg, config.getRelationType());
-        withCallback(processEntityRelationAction(ctx, msg, relationType),
-                filterResult -> ctx.tellNext(filterResult.getMsg(), filterResult.isResult() ? SUCCESS : FAILURE), t -> ctx.tellFailure(msg, t), ctx.getDbCallbackExecutor());
+        withCallback(processEntityRelationAction(ctx, msg),
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/action/TbAbstractRelationActionNode.java:70]: replace processEntityRelationAction with onMsg in each sub class, to remove TbPair usages.

**Review by ViacheslavKlimov — APPROVED** [review]

**Pull request merged**